### DISCUSSION TITLE
Lossless smart cut via image sequences?

### DISCUSSION THREAD

**Discussion by ghost** [mifi/lossless-cut#1405] [Ideas]
As it is currently implemented, smart cut converts needed fragment between keyframes into H265(?), which in my experience is noticeably lossy.

However there is a better way. Get the needed segment in between keyframes, export all frames as lossless pictures, get the cut frames and assemble them back into video. Merge resulted video with the regular keyframe sections.

Is there something I'm missing? Pretty sure Lossless cut alredy supports exporting to pictures.

**Comment by mifi** [comment]
1. What exactly do you mean by "export all frames as lossless pictures"? You mean export each frame as PNG?
2. And how to assemble PNGs together into a video?
3. How to merge this video with the H265 video? AFAIK APNG video cannot be combined with H265 in the same stream.

**Comment by ghost** [comment]
Yeah I think I overlooked the merging back process a bit, assuming it would be fine. 😬

> What exactly do you mean by "export all frames as lossless pictures"? You mean export each frame as PNG?

Yes exactly

> And how to assemble PNGs together into a video?

Well here is the command that I used:

`ffmpeg -framerate 25 -i '*.png' -c:v libx264rgb -crf 0 out.mkv`

This does produce a visually lossless H264 video, but...

> How to merge this video with the H265 video? AFAIK APNG video cannot be combined with H265 in the same stream.

Not quite sure how H265 would work as the original video I wanted to cut was H264, so that's what I went with.
I tried to merge both videos, picture and original, via:

`ffmpeg -f concat -safe 0 -i files.txt -codec copy output.mkv`

And it kinda worked. I did get a complete file, but ffmpeg threw this message my way: 

`Auto-inserting h264_mp4toannexb bitstream filterA`

Which puts a weird brightness(?) filter over the original video. Couldn't find a whole lot about it. I think ffmpeg uses it when 2 H264 files use different pixel format? That was the only difference I found between files, original being `yuv420p` and picture  `gbrp`. I tried to pass `-pix_fmt yuv420p` when assembling the picture video but ffmpeg just ignored it.

**Reply by mifi** [reply]
> Well here is the command that I used:
> `ffmpeg -framerate 25 -i '*.png' -c:v libx264rgb -crf 0 out.mkv`
> 
> This does produce a visually lossless H264 video, but...

This is basically what losslesscut is already doing. It is not lossless. What losslesscut could do is provide a way for the user to change the quality of the encoded part (even visually lossless quality) - see recent discussions about this here #126
